### ISSUE TITLE
Getting record_overflow(22) error in TLS1.3 mode

### ISSUE THREAD

**Issue by alluettiv** [Mbed-TLS/mbedtls#8758]
### Summary
Server app is using mbedTLS 3.5.2. Client app is using BouncyCastle 2.2.1 .NET library.
When sending large files, mbedTLS sends it in chunks of 16416 bytes. Client SSL library states that this size is invalid for given ciphersuite, and should be 16401 bytes at most for TLS/AEAD.

### System information
Mbed TLS version 3.5.2

### Debug info from mbedTLS
ssl_msg.c:3032: 0x7efcc8002138: output record: msgtype = 23, version = [3:3], **msglen = 16416**
ssl_msg.c:3037: 0x7efcc8002138: dumping 'output record sent to network' (**16421 bytes**)
ssl_msg.c:3037: 0x7efcc8002138: 0000:  17 03 03 40 20 18 ed 56 53 49 8a df 7a af 3d 67  ...@ ..VSI..z.=g

### Debug info from BouncyCastle
Server version: TLS 1.3
Selected Cipher Suite: 1302
Server certificate: count=2
Server certificate: alg=1.2.840.113549.1.1.11

```
public virtual int GetCiphertextDecodeLimit(int plaintextLimit)
{
    int innerPlaintextLimit = plaintextLimit + (m_decryptUseInnerPlaintext ? 1 : 0); // 16384 + 1
    return innerPlaintextLimit + m_macSize + m_record_iv_length;  // 16385 + 16(macSize) + 0(iv_len)
}
```
So GetCiphertextDecodeLimit() returns 16401 for TlsAead as maximum allowed record size. So when mbedTLS sends a block larger than that the exception is thrown by BouncyCastle and connection is aborted.

**Comment by paul-elliott-arm** [comment]
Hi, would it be possible to attach your mbedtls_config.h?

**Comment by alluettiv** [comment]
Hi, config attached. But I think it's config-independent. From what I've seen in the debugger, it works like this:
1. **ssl_write_real()** is called with actual big buffer (like 2MB or so).
2. **mbedtls_ssl_get_max_out_record_payload()** returns **16384**, then it proceeds to **mbedtls_ssl_write_record()**
3. **mbedtls_ssl_encrypt_buf()** calls **ssl_compute_padding_length(16384, 1)** which returns **16340** - it pads the buffer to nearest 16 byte boundary (MBEDTLS_SSL_CID_TLS1_3_PADDING_GRANULARITY) - is this behavior conforms to TLS standard?
4. **ssl_build_inner_plaintext()** writes content type byte, then zero-fills 15 bytes of padding, and then updates content length from **16384** to **16340**.

I did not look further, but I presume those extra 16 bytes in **16416** total to be sent is AEAD tag.

So basically extra 15 bytes of the padding added everywhere is what creates the problem on client size. Is it allowed for record to be larger than 16384+1+16 bytes for AEAD according to standards? Or perhaps mbedtls_ssl_get_max_out_record_payload should return 16383 to accommodate extra recordType byte, so it won't cause padding to be added? If not, is padding actually needed? Since BouncyCatle TLS implementation expects maximum to be of 16401 bytes which implies 16384+1+16 and no padding.

[mbedtls_config.zip](https://github.com/Mbed-TLS/mbedtls/files/14087323/mbedtls_config.zip)

**Comment by mpg** [comment]
Let's check what [RFC 8446 Section 5](https://datatracker.ietf.org/doc/html/rfc8446#section-5) has to say about this. There it is, [section 5.4](https://datatracker.ietf.org/doc/html/rfc8446#section-5.4):

>   The presence of padding does not change the overall record size
>   limitations: the full encoded TLSInnerPlaintext MUST NOT exceed 2^14
>   + 1 octets.

So it looks like we're clearly in the wrong by adding padding here. I think we want to fix `ssl_compute_padding_length()` in order to make it aware of this hard limit on padded size.

For reference, `TLSInnerPlaintext` was previously defined as:

>       struct {
>           opaque content[TLSPlaintext.length];
>           ContentType type;
>           uint8 zeros[length_of_padding];
>       } TLSInnerPlaintext;

and `TLSPlaintext.length` is the `len` argument passed to `ssl_compute_padding_length()`.

Since `ssl_compute_padding_length()` is also used for DTLS 1.2 + CID, we probably want to check if the same limitation applies there. Ah, that's easy: [RFC 9146 Section 4](https://datatracker.ietf.org/doc/html/rfc9146#name-record-layer-extensions) just refers to RFC 8446 Section 5.4 regarding padding. So, we can keep the same function for both then.

**Comment by mpg** [comment]
Looks easy enough to fix and to test: just send a max-sized packed (we should hopefully already have tests doing this) and check (for example from the debug output) that the size of the encrypted record is as expected. (We can force a ciphersuite in the test in order to know for sure what the AEAD expansion is.) So, labeling size-s.

**Comment by ronald-cron-arm** [comment]
This has probably been fixed in development as part of the work on record size limit. We have had similar issues: #8325 and #7918. I am not sure if we a non-regression test yet though. @waleed-elmelegy-arm, could you have a look?

**Comment by waleed-elmelegy-arm** [comment]
@alluettiv  Can you check with the latest development version? like Ronal said a problem like this have been solved while working on the record size limit, Thanks.

**Comment by alluettiv** [comment]
@waleed-elmelegy-arm I tried development version - seems to work fine. I didn't debug it, but in the same scenario the problem is not reproducible anymore.

**Comment by tom-cosgrove-arm** [comment]
@alluettiv Thanks for confirming. I'm closing this issue as fixed in development
